src/main/java/dev/morling/onebrc/CalculateAverage_royvanrijn.java

```diff
@@ -250,72 +234,105 @@ private static long compilePattern(final byte value) {
+        /**
+         * Separated hashtable, keeps memory local, idea of from franz1981.
+         * And colocate measurements in Entry:
+         */
+        private final int[] hashTable = new int[TABLE_SIZE];
+
+        static final class Entry {
+            private final long[] data;
+            private final String city;
+            private int min, max, count;
+            private long sum;
+
+            Entry(long[] data, String city) {
+                this.data = data;
+                this.city = city;
+                this.min = 1000;
+                this.max = -1000;
+            }
 
-        private MeasurementRepository.Entry[] table = new MeasurementRepository.Entry[tableSize];
+            public void updateWith(int measurement) {
+                min = min(min, measurement);
+                max = max(max, measurement);
+                sum += measurement;
+                count++;
+            }
 
-        record Entry(long address, long[] data, int length, int hash, String city, Measurement measurement) {
+            public Entry mergeWith(Entry entry) {
+                min = min(min, entry.min);
+                max = max(max, entry.max);
+                sum += entry.sum;
+                count += entry.count;
+                return this;
+            }
 
-            @Override
             public String toString() {
-                return city + "=" + measurement;
+                return round(min) + "/" + round((1.0 * sum) / count) + "/" + round(max);
+            }
+
+            private static double round(double value) {
+                return Math.round(value) / 10.0;
             }
         }
 
-        public void update(long address, long[] data, int length, int hash, int temperature) {
+        public void update(final long address, final long[] data, final int length, final int hash, final int temperature) {
 
-            int dataLength = length >> 3;
-            int index = hash & tableMask;
-            MeasurementRepository.Entry tableEntry;
-            while ((tableEntry = table[index]) != null
-                    && (tableEntry.hash != hash || tableEntry.length != length || !arrayEquals(tableEntry.data, data, dataLength))) { // search for the right spot
-                index = (index + 1) & tableMask;
-            }
+            final int dataLength = (length >> 3) + 1;
 
-            if (tableEntry != null) {
-                tableEntry.measurement.updateWith(temperature);
-                return;
-            }
+            int index = hash & TABLE_MASK;
+            Entry tableEntry = null;
 
-            // --- This is a brand new entry, insert into the hashtable and do the extra calculations (once!) do slower calculations here.
-            Measurement measurement = new Measurement();
+            // Find the entry:
+            while (true) {
+                int hashTableEntry;
+                if ((hashTableEntry = hashTable[index]) == 0) {
+                    // Slot is empty
```

Comment:
> I think your current solution has one cache miss to access hashTable[index], another for table[index]

these are there on purpose: Intel uses LineFill buffers which can be used so resolve dependent loads in parallel.
The reason by splitting them in two is to leverage such; so you have 2 cache misses, but only the slower will dominate the cost.
In addition using a hash table make it compact enough that (likely?) subsequent accesses can still find it hot, while entries, being bigger, maybe not.

> You could make the long[] a bit larger to store the hashCode as the first element.

I've tried it already, before proposing this solution, and I get a performance hit if compared to what we have now.

Reply:
and maybe simply remove the hashCode comparison?
Then we delve in optimizing for the benchmark, as the test file has a few hundreds of entries, and if no collisions, we can get away with skipping this check

Reply:
That is a good point!

Reply:
@royvanrijn

Reply:
@obourgain beware that the one used by Gunnar should have 10K iirc

Reply:
he said that the code must support 10k, and there is a test for that. I don't know if he generated a new test file with more data, but the initial one was probably generated by the first version of the generator

Reply:
You still have to compute the hashCode, but you may not store it

Reply:
@gunnarmorling what are you using to generate the score table? Which file?

Reply:
@obourgain I’ve tried at least a couple of versions of that, just keeping the long[]. What I might try to implement is a store more like… a prefix tree? Where if a certain first long as data hasn’t been seen, write as new entry, if it has, continue until you don’t match or completely match. Will need to look up, I think a similar/suitable data structure exists.

Reply:
Thanks @obourgain for your analysis, which made me redo the test with the larger data set (obtained with create measurements 3) vs the default one and with the default, probably because the table size is way smaller (just 100 entries!), the additional in-flight load for a separate hash table is damaging performance, so... @royvanrijn to summarize:
- very big entry table -> separate hash table of hash bring advantages
- very small entry table -> separate hash table reduce performance (my box goes from 1.8 -> 1.7 secs)